### ISSUE TITLE
enable=all enables shell-specific checks that contradict shell=sh

### ISSUE THREAD

**Issue by peturingi** [koalaman/shellcheck#3399]
`enable=all` enables shell-specific checks when used together with `shell=sh`.
This defeats the purpose of `shell=sh` when used with `enable=all` as users must refer to [The Open Group Base Specification](https://pubs.opengroup.org/onlinepubs/9799919799/) for each warning to determine whether or not to disable _false positive_ warnings.

#### For bugs with existing features
- Rule Id (if any, e.g. SC1000): SC2292 (and potentially others)
- My shellcheck version (`shellcheck --version` or "online"): 0.11.0
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2292)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:
```sh
#!/bin/sh
# shellcheck shell=sh enable=all

if [ "$foo" = "bar" ]; then
    echo yes
fi
```

#### Here's what shellcheck currently says:
```
In line 5:
if [ "$foo" = "bar" ]; then
   ^-- SC2292 (style): Prefer [[ ]] over [ ] for tests in Bash/Ksh.
```

#### Here's what I wanted or expected to see:

No warning. `[[ ]]` is not POSIX, and I have explicitly requested POSIX compliance with `shell=sh`.

`enable=all` should respect the `shell=` directive and only enable checks appropriate for the target shell. Checks recommending non-POSIX constructs should be automatically excluded when `shell=sh`.

More generally: optional checks should be tagged with their applicable shells, and `enable=all` should be interpreted as "enable all checks valid for the specified shell".

#### Workaround
```sh
# shellcheck shell=sh enable=all disable=SC2292
```
